frontend/src/components/course-search/CourseSearch.tsx

```diff
@@ -87,25 +92,74 @@ export default memo(function CourseSearch() {
         conflictingSectionsOptions,
     ]);
 
+    const { enableHistoricalSearch, historicalSearchRange } = useStore(
+        (store) => store.experimentalFeaturesOptions,
+    );
+    const allTerms = useAllTerms() ?? [];
+    const range = Math.min(allTerms.length, historicalSearchRange);
+    const historicalSections = useSectionsForTermsQuery(
+        enableHistoricalSearch && sectionsToShow?.length === 0,
+        allTerms.slice(0, range),
+    ).data;
+
+    const matchingHistoricalSections: APIv4.Section[] | undefined =
+        React.useMemo(() => {
```

Comment:
you are performing the entire historical search here whether the result is needed or not and is probably causing the performance issue

Reply:
I will fix this by implement if-check that prevent the "expensive" course searching algorithm by returning `undefined` when the data is not needed (done easily) --> this should remove any unnecessary computation down to a simple check while we keep using the useMemo to get the benefit of simple memoization.

I will also look into the hooks you recommend, and see if they yield better performance.